docs/architecture.md

```diff
@@ -0,0 +1,252 @@
+
+A waypoint marker floating over an objective has a graphical appearance and a position in space, yet is not affected by physics.
+A trigger which detects that the player has reached the objective is nearly identical, but is invisible and thus lacks a graphical appearance.
+
+This broad spectrum of combinations of behaviors indicates that game objects are best represented by **composition**.
+Rather than attempting to force game objects into a hierarchy of behaviors (e.g. using inheritance to organize "universal" properties that we *assume* will be shared), we would like to make no assumptions, and instead freely attach any behvaior to any object.
+
+Furthermore, we would like to **decouple** systems from one another, wherever possible.
+The physics logic and rendering logic for an object both may rely on the object's position to do their job, but they need not know any more.
+Decoupling unrelated systems from one another keeps them focused, and easier to work with.
+
+### Solution
+
+These problems are addressed by the **Entity Component System** (ECS) paradigm. Here's how it works:
+
+Everything in your game is an **Entity**. The player is an entity, the terrain is an entity, a projectile in flight is an entity, and so on.
+Under the hood, an entity is represented by nothing more than a unique integer ID.
+
+We may attach **Components** to entities.
+A component represents some behavior or property that the entity possesses, such as a "Position," "Color," "Health," and so on.
+Under the hood, each component type is a `struct` which stores the state data relevant to its function (e.g. a 3D vector representing position, a mesh resource to be drawn, etc.).
+
+**Systems** are responsible for providing the behaviors promised by the components. Systems only process entities which possess the relevant components.
+A physics system would process all entities which a "position" and "velocity" component.
+A graphics system might operate on the "position" and a "mesh" component.
+Entities with a "Mass" component would be handled by the "Gravity" system.
+Any entities which don't possess the right components are simply ignored, and for the entities which *are* processed, any components unrelated to the system are ignored.
+
+This arrangement provides us the decoupling we desired; if we want to add a new behavior to a game object, say, to make it flash colors, we simply create a new function that accesses the "Color" component of any entity which has one, and modifies its color. No other information about the entity is relevant, and the presence of any one component doesn't imply anything about the entity beyond that component's narrow purpose.
+
+## ECS in OSP
+
+In OSP, ECS powers the `Universe` and `ActiveScene` using the [EnTT](https://github.com/skypjack/entt/) library.
+
+Components **never inherit each other**, and all common properties are split into separate components.
```

Comment:
This applies to ECS in general, and not just OSP. One additional thing to note is that inheriting 'components' is very common in mainstream game engines like Unity and Unreal.

Reply:
I was torn over where to put that but I was trying to keep any implementation-related details out of the overview section. If you like the rest, I could make a brief "ECS implementation" section that covers more general aspects of ECS implementations (contiguous storage, no inheritance, static functions). That would be a good place to highlight how ECS plays very nicely with data-oriented design.

Reply:
Hmm, I kind of thought this section was a heading 3. Pretty much everything I've typed in the ECS section is already general. ECS implementations just happen to vary quite a bit; this whole section can just be "ECS implementation."